ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,412 @@
+
+```ruby
+interface _PayloadConverter[U]
+  # Takes any Ruby object and returns a Payload object
+  def to_payload: (U) -> Temporal::Payload
+
+  # Takes a Payload object and returns a Ruby object
+  def from_payload: (Temporal::Payload) -> U
+end
+```
+
+The SDK will provide an implementation of basic converters — Null (for `nil`s), Bytes (any stream of
+bytes) and JSON (using the `oj` gem and potentially allowing the SDK user to specify the
+[mode](https://github.com/ohler55/oj/blob/develop/pages/Modes.md)). We will also provide a
+`Composite` converter to combine multiple converters behind the same interface.
+
+In most cases the SDK users are expected to use the `Composite` converter and add their custom
+converters to it. These are expected to have a slightly different interface:
+
+```ruby
+interface _EncodingPayloadConverter[U]
+  # Returns a MIME type that this converter provides
+  def encoding: -> String
+
+  # Takes any Ruby object and returns a Payload object or nil
+  # A nil response indicates that this converter was unable to convert the given value
+  def to_payload: (U) -> Temporal::Payload?
+
+  # Takes a Payload object and returns a Ruby object
+  def from_payload: (Temporal::Payload) -> U
+end
+```
+
+While a top-level converter is expected to convert any value to a Payload and back (or throw an
+error otherwise), an encoding payload converter (the one used with the `Composite` converter) is
+only expected to handle values and Payload matching it's specified encoding.
+
+
+#### JSON Converter
+
+The default JSON converter will respect [the interface](https://ruby-doc.org/stdlib-3.1.2/libdoc/json/rdoc/JSON.html#module-JSON-label-Custom+JSON+Additions)
+provided by the Ruby's JSON module to serialize/deserialize an object.
+
+
+### Codecs
+
+And a payload codec is expected to implement this interface:
+
+```ruby
+interface _PayloadCodec
+  # Takes an array of Payload objects and returns an array of encoded Payload object
+  # The resulting Array is expected to hold at least one Payload
+  def encode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+
+  # Takes an array of Payloads object and returns an array of decoded Payload object
+  # The resulting Array is expected to hold at least one Payload
+  def decode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
+end
+```
+
+
+## Errors
+
+Here's the Error hierarchy that is planned for the Ruby SDK:
+
+```ruby
+module Temporal
+  # top level error superclass
+  class Error < StandardError
+
+  # superclass for all errors within the SDK itself
+  class InternalError < Error
+
+  # errors specific to Client
+  class ClientError < Error
+
+  # errors specific to Worker
+  class WorkerError < InternalError
+
+  # superclass for connection, network and response errors
+  class RPCError < Error
+  class UnexpectedError < RPCError
+
+  # superclass for all API failure responses
+  class APIError < ServerError
```

Comment:
I think this should be called "Failure" or similar since it is basically 1:1 with "Temporal Failures" and it's not always "API" related (at least not obviously to the user which doesn't use our API). Also, it should carry the raw proto with it probably.

Reply:
It does if you are talking about consistency across APIs. We don't embed the word API hardly anywhere. They don't know that API is even involved.

Reply:
Well, this is not true, since all the protobufs clearly define a `Temporal::Api`. So both `Failure` and `Api` have the same source and are never shown to the user.

But I'm not too fussy about it, happy to call it `Failure` if you think it's clearer (I don't)

Reply:
One of the primary reasons for these errors is to hide/convert the protobufs. I don't mind something besides "Failure", I just think "API" is not best. I choose "Failure" because it aligns with other SDKs and docs and there is no value in uniqueness here.

Reply:
I find it weird that ApplicationError subclasses ServerError.
I would change the hierarchy and possibly rename APIError to something like FailureError.
The alternative (which @cretz will disagree with) is to call the failures base class TemporalFailure and suffix its subclasses with Failure like Java and TS do.
I think it's nice to cleanly mark and group all of these failure types this way.

Reply:
> The alternative (which @cretz will disagree with) is to call the failures base class TemporalFailure and suffix its subclasses with Failure like Java and TS do.

I don't disagree with this here. I only disagree in Python where errors are suffixed with `Error`. I'm all for a common suffix on these and If unless it's expected to always have the `Error` suffix, I think `Failure` works.

Reply:
I've renamed the error to `FailureError` and made it a direct subclass of `Temporal::Error` 👍